src/interactions/panZoomInteraction.ts

```diff
@@ -144,14 +166,41 @@ export module Interactions {
       });
     }
 
-    private static _magnifyScale<D>(scale: QuantitativeScale<D>, magnifyAmount: number, centerValue: number) {
+    private _magnifyScale<D>(scale: QuantitativeScale<D>, magnifyAmount: number, centerValue: number) {
       var magnifyTransform = (rangeValue: number) => scale.invert(centerValue - (centerValue - rangeValue) * magnifyAmount);
-      scale.domain(scale.range().map(magnifyTransform));
+      var transformedDomain = scale.range().map(magnifyTransform);
+      var constrainedDomain = this._constrainedDomain(scale, transformedDomain);
+      scale.domain(<any[]> constrainedDomain);
     }
 
-    private static _translateScale<D>(scale: QuantitativeScale<D>, translateAmount: number) {
+    private _translateScale<D>(scale: QuantitativeScale<D>, translateAmount: number) {
       var translateTransform = (rangeValue: number) => scale.invert(rangeValue + translateAmount);
-      scale.domain(scale.range().map(translateTransform));
+      var transformedDomain = scale.range().map(translateTransform);
+      var constrainedDomain = this._constrainedDomain(scale, transformedDomain);
+      scale.domain(<any[]> constrainedDomain);
+    }
+
+    private _constrainedDomain<D>(scale: QuantitativeScale<D>, domainToConstrain: D[]) {
+      var domainValues = <number[]> domainToConstrain.map((d) => d.valueOf());
+      var domainExtent = Math.abs(domainValues[1] - domainValues[0]);
+      var domainCenter = (domainValues[1] + domainValues[0]) / 2;
+
+      var constrainedDomainValues = [Math.min(domainValues[0], domainValues[1]), Math.max(domainValues[0], domainValues[1])];
+      var minDomainExtent = this._minDomainExtents.get(scale);
+      if (domainExtent < minDomainExtent) {
+        constrainedDomainValues[0] = domainCenter - minDomainExtent / 2;
+        constrainedDomainValues[1] = domainCenter + minDomainExtent / 2;
+      }
+
+      var maxDomainExtent = this._maxDomainExtents.get(scale);
+      if (domainExtent > maxDomainExtent) {
+        constrainedDomainValues[0] = domainCenter - maxDomainExtent / 2;
+        constrainedDomainValues[1] = domainCenter + maxDomainExtent / 2;
+      }
+
+      var constrainedDomain = constrainedDomainValues.map((value) => scale instanceof Scales.Time ? new Date(value) : value);
```

Comment:
I don't think we should have an `instanceof` conditional here, which suggests that this might not be quite the right implementation. Should the constraint logic live on the `Scale`, with the actual constraint values stored on the `PanZoom`?